transports/quic/src/transport.rs

```diff
@@ -654,6 +728,51 @@ pub(crate) enum SocketFamily {
     Ipv6,
 }
 
+#[derive(Clone)]
+pub(crate) enum ConnectingMode {
+    QUIC,
+    WebTransport(Vec<Multihash<64>>, libp2p_noise::Config),
+    Mixed(Vec<Multihash<64>>, libp2p_noise::Config),
+}
+
+impl ConnectingMode {
+    pub(crate) fn update_multiaddr(&self, addr: Multiaddr) -> Multiaddr {
+        fn add_web_transport_protocol(v: &Vec<Multihash<64>>, addr: Multiaddr) -> Multiaddr {
+            let mut vec = v.clone();
+            let mut res = addr.with(Protocol::WebTransport).with(Protocol::Certhash(
+                vec.pop().expect("Gets the last element"),
+            ));
+            if !vec.is_empty() {
+                res = res.with(Protocol::Certhash(
+                    vec.pop().expect("Gets the last element"),
+                ));
+            };
+
+            res
+        }
+
+        match self {
+            ConnectingMode::QUIC => addr,
+            ConnectingMode::WebTransport(hashes, _) => add_web_transport_protocol(hashes, addr),
+            ConnectingMode::Mixed(hashes, _) => add_web_transport_protocol(hashes, addr),
```

Comment:
Don't think we need to extract the logic into `add_web_transport_protocol` if we just have both variants on the same branch.

```suggestion
            ConnectingMode::QUIC => addr,
            ConnectingMode::WebTransport(hashes, _) |  ConnectingMode::Mixed(hashes, _) => add_web_transport_protocol(hashes, addr),
```